### DISCUSSION TITLE
Path wrapper type

### DISCUSSION THREAD

**Discussion by remimimimimi** [vlang/v#12692] [Ideas]
### Motivation

It was after working on a number of projects that use file paths in various ways. System-filepath library, will solve many path problems by being an opaque path type. It occurred to me that the same kind of bugs kept cropping up:

  - Expected a path to be absolute but it was relative, or vice-versa.

  - Expected two equivalent paths to be equal or order the same, but they did not (/home//foo vs /home/foo/ vs /home/bar/../foo, etc.).

  - Unpredictable behavior with regards to concatenating paths.

  - Confusing files and directories.

  - Not knowing whether a path was a file or directory or relative or absolute based on the type alone was a drag.

All of these bugs are preventable.

### Approach

My approach to problems like this is to make a type that encodes the properties I want and then make it impossible to let those invariants be broken, without compromise or backdoors to let the wrong value “slip in”. Once I have a path, I want to be able to trust it fully. This theme will be seen throughout the things I lay out below.

### Possible Solution

After having to fix bugs due to these, I came to the following points:

  - An opaque Path type (a newtype wrapper around string).

  - Smart constructors which are very stringent in the parsing.

  - Make the parsers highly normalizing.

  - Leave equality and concatenation to basic string equality and concatenation.

  - Include relativity (absolute/relative) and type (directory/file) in the type itself.

  - Use the already cross-platform filepath package for implementation details.

**Comment by larpon** [comment]
How should it treat:
* Symbolic links across platforms?
* Hard links?
* Network drives?
* Embedded files?
* Unix mount points?
* Windows drive prefixes?

Qt has this huge level of abstraction that also covers URLs and embedded qrc resources - and it has very little benefits, a lot of pitfalls you need to workaround and takes a lot of doc reading to understand.

I'm not sure a Path type and all the extra overhead and complexity is worth it?

Personally I've come a long way with:
1. `os.join_path()`
2. `os.real_path()`
3. `'string'.trim_right('/')`
4. `os.path_separator`

In the above order from "most used" to "least used"

**Comment by JalonSolov** [comment]
`os.real_path()` takes care of a lot of the proposed changes already, since it will resolve all the `../` relative path stuff.

Of course, the problem there is the problem with Zip file paths, where you can construct a path with relative bits that winds up trying to overwrite protected areas.  As long as that's guarded against...

**Comment by remimimimimi** [comment]
I think that we should just create a wrapper around the string which will hide all these functions calls inside it. 
It'll have several pros over just accepting random string:
 - Foolproof, this is rude but in my opinion very short description of a problem. In other words, it will help programmers rely on type checks of the compiler instead of thinking about all this stuff by them.
 - From the preceding paragraph follows an increase in maintainability.
 - And also because everywhere where we'll use paths it needs to be explicit, it also increases readability. Because there's will be Path type, instead of random string about which we have no context.
 - Wrapper give us the possibility of changing API of path normalization.
 - `Path` constructor will take care of path validation so @JalonSolov problem can also be solved inside it.

And there are only one small visible cons about this wrapper:
 - Wrapping and unwrapping, which I don't see a big problem taking into account the leading advantages.

> Windows drive prefixes?

This can be solved by `root()` method which will return root of the path (e.g. `'/'` on UNIX and drive name on windows).

I do not know yet about other points, but if they are already being solved by the current API, it won't be a problem to incapsulate inside `Path` type, I suppose.

Most of the argumentation I take from [this Haskell package](https://hackage.haskell.org/package/path). Didn't read it fully but if you want more answers and inspiration you can read through it already, as well as from the [rust std library](https://doc.rust-lang.org/std/path/index.html) which also has Path type but a much more complicated one.
